cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkTransactionService.java

```diff
@@ -0,0 +1,313 @@
+   */
+  private void jobStarted(JobTransaction jobTransaction) {
+    Integer jobId = jobTransaction.getJobId();
+
+    // Remember the job Id. We won't start a new transaction here until a stage requested for it.
+    // This is because there can be job that doesn't need transaction or explicit transaction is being used
+    JobTransaction existingJobTx = jobTransactions.putIfAbsent(jobId, jobTransaction);
+    if (existingJobTx != null) {
+      // Shouldn't happen as Spark generates unique Job Id.
+      // If that really happen, just log and return
+      LOG.error("Job already running: {}", existingJobTx);
+      return;
+    }
+
+    // Build the stageId => jobId map first instead of putting to the concurrent map one by one
+    Map<Integer, Integer> stageToJob = new HashMap<>();
+    for (Integer stageId : jobTransaction.getStageIds()) {
+      stageToJob.put(stageId, jobId);
+    }
+    this.stageToJob.putAll(stageToJob);
+  }
+
+  /**
+   * Notifies the given job execution completed.
+   *
+   * @param jobId the unique id that identifies the job.
+   * @param succeeded {@code true} if the job execution completed successfully.
+   */
+  void jobEnded(Integer jobId, boolean succeeded) {
+    JobTransaction jobTransaction = jobTransactions.remove(jobId);
+    if (jobTransaction == null) {
+      // Shouldn't happen, otherwise something very wrong. Can't do much, just log and return
+      LOG.error("Transaction for job {} not found.", jobId);
+      return;
+    }
+
+    // Cleanup the stage to job map
+    stageToJob.keySet().removeAll(jobTransaction.getStageIds());
+
+    // Complete the transaction
+    jobTransaction.completed(succeeded);
+  }
+
+  /**
+   * HTTP Handler to provide the Spark stage execution transaction lookup service.
+   */
+  public final class SparkTransactionHandler extends AbstractHttpHandler {
+
+    private final TransactionCodec txCodec = new TransactionCodec();
+
+    /**
+     * Handler method to get a serialized {@link Transaction} for the given stage.
+     */
+    @GET
+    @Path("/spark/stages/{stage}/transaction")
+    public void getTransaction(HttpRequest request, HttpResponder responder, @PathParam("stage") int stageId) {
+      // Lookup the jobId from the stageId
+      Integer jobId = stageToJob.get(stageId);
+      if (jobId == null) {
+        // If the JobId is not there, it's either the job hasn't been registered yet (because it's async) or
+        // the job is already finished. For either case, return 404 and let the client to handle retry if necessary.
+        responder.sendString(HttpResponseStatus.NOT_FOUND, "JobId not found for stage " + stageId);
+        return;
+      }
+
+      // Get the transaction
+      JobTransaction jobTransaction = jobTransactions.get(jobId);
+      if (jobTransaction == null) {
+        // The only reason we can find the jobId from the stageToJob map but not the job transaction is because
```

Comment:
`jobTransaction` can also be `null` in case if the the job itself does not need any transaction(not doing any read/write operation on the datasets), or this endpoint is called from a stage which is before the stage in the job which requires a transaction, right? In that case we might want to update the comment here.